Add a default implementation of __torch_dispatch__

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73727
* __->__ #73684

I was working on an explanation of how to call into the "super"
implementation of some given ATen operation inside of __torch_dispatch__
(https://github.com/albanD/subclass_zoo/blob/main/trivial_tensors.py)
and I kept thinking to myself "Why doesn't just calling super() on
__torch_dispatch__ work"?  Well, after this patch, it does!  The idea
is if you don't actually unwrap the input tensors, you can call
super().__torch_dispatch__ to get at the original behavior.

Internally, this is implemented by disabling PythonKey and then
redispatching.  This implementation of disabled_torch_dispatch is
not /quite/ right, and some reasons why are commented in the code.
There is then some extra work I have to do to make sure we recognize
disabled_torch_dispatch as the "default" implementation (so we don't
start slapping PythonKey on all tensors, including base Tensors),
which is modeled the same way as how disabled_torch_function is done.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D34590357](https://our.internmc.facebook.com/intern/diff/D34590357)

ezyang committed:
Add a default implementation of __torch_dispatch__

I was working on an explanation of how to call into the "super"
implementation of some given ATen operation inside of __torch_dispatch__
(https://github.com/albanD/subclass_zoo/blob/main/trivial_tensors.py)
and I kept thinking to myself "Why doesn't just calling super() on
__torch_dispatch__ work"?  Well, after this patch, it does!  The idea
is if you don't actually unwrap the input tensors, you can call
super().__torch_dispatch__ to get at the original behavior.

Internally, this is implemented by disabling PythonKey and then
redispatching.  This implementation of disabled_torch_dispatch is
not /quite/ right, and some reasons why are commented in the code.
There is then some extra work I have to do to make sure we recognize
disabled_torch_dispatch as the "default" implementation (so we don't
start slapping PythonKey on all tensors, including base Tensors),
which is modeled the same way as how disabled_torch_function is done.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34590357).

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -151,10 +151,12 @@ static const char* VOLATILE_WARNING =
 
 static bool check_has_torch_dispatch(PyObject *obj) {
   PyTypeObject *tp = Py_TYPE(obj);
+  auto* attr = PyObject_FastGetAttrString(obj, "__torch_dispatch__").ptr();
```
TIL: we have that function.

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -151,10 +151,12 @@ static const char* VOLATILE_WARNING =
 
 static bool check_has_torch_dispatch(PyObject *obj) {
   PyTypeObject *tp = Py_TYPE(obj);
+  auto* attr = PyObject_FastGetAttrString(obj, "__torch_dispatch__").ptr();
```
Not sure what the auto does here but this returns a new reference, so we actually own it.

ezyang replied:
Ach, I see, the ptr() idiom is a bit dangerous but it works out because the PyObject in question has lifetime independent of PyObject_FastGetAttrString. (Usually.) Can fix anyway. There are a bunch of other sites that need to get similarly fixed too.

albanD replied:
> the PyObject in question has lifetime independent of PyObject_FastGetAttrString

What do you mean?
If this function has the same semantic as python's `PyObject_GetAttrString` then you get an owning reference that you have to decref here!

ezyang replied:
But we return a py::object so explicit decref is not needed, the class handles it. But since it's a temporary in 154 I will immediately decref at the end of 154 even though I still have a raw pointer. In any case I have fixed it.

albanD commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -127,6 +136,43 @@ PyObject* THPModule_disable_torch_function(PyObject *self, PyObject *a) {
   END_HANDLE_TH_ERRORS
 }
 
+PyObject* THPModule_disable_torch_dispatch(PyObject *self, PyObject *a) {
```
Not sure why you need an implementation here (and above for torch function)? Shouldn't these never be called?

ezyang replied:
They do get called if you `super().__torch_dispatch__` !

albanD replied:
Ho ok!
Sound good!

albanD reviewed (commented):
Sounds quite good!

bdhirsh commented on test/test_python_dispatch.py:
```diff
@@ -641,6 +641,52 @@ def test_autograd_in_attr(self):
+            def __new__(cls, elem):
+                return torch.Tensor._make_subclass(cls, elem)
+
+            __torch_function__ = torch._C._disabled_torch_function_impl
```
question unrelated to this PR - right now if you want to "just use torch_dispatch", you need to manually disable torch_function like the above.

It seems like using both of them is the exception rather than the common case. Do you think we'd wanna consider flipping the default so that torch_function is default disabled for your subclass if you write your own torch_dispatch? (Open question around how BC-breaking that is)

ezyang replied:
We should do this, and I think literally no one will be BC broken by it.

ezyang committed:
Update on "Add a default implementation of __torch_dispatch__"


I was working on an explanation of how to call into the "super"
implementation of some given ATen operation inside of __torch_dispatch__
(https://github.com/albanD/subclass_zoo/blob/main/trivial_tensors.py)
and I kept thinking to myself "Why doesn't just calling super() on
__torch_dispatch__ work"?  Well, after this patch, it does!  The idea
is if you don't actually unwrap the input tensors, you can call
super().__torch_dispatch__ to get at the original behavior.

Internally, this is implemented by disabling PythonKey and then
redispatching.  This implementation of disabled_torch_dispatch is
not /quite/ right, and some reasons why are commented in the code.
There is then some extra work I have to do to make sure we recognize
disabled_torch_dispatch as the "default" implementation (so we don't
start slapping PythonKey on all tensors, including base Tensors),
which is modeled the same way as how disabled_torch_function is done.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D34590357](https://our.internmc.facebook.com/intern/diff/D34590357)

[ghstack-poisoned]

ezyang committed:
Update on "Add a default implementation of __torch_dispatch__"


I was working on an explanation of how to call into the "super"
implementation of some given ATen operation inside of __torch_dispatch__
(https://github.com/albanD/subclass_zoo/blob/main/trivial_tensors.py)
and I kept thinking to myself "Why doesn't just calling super() on
__torch_dispatch__ work"?  Well, after this patch, it does!  The idea
is if you don't actually unwrap the input tensors, you can call
super().__torch_dispatch__ to get at the original behavior.

Internally, this is implemented by disabling PythonKey and then
redispatching.  This implementation of disabled_torch_dispatch is
not /quite/ right, and some reasons why are commented in the code.
There is then some extra work I have to do to make sure we recognize
disabled_torch_dispatch as the "default" implementation (so we don't
start slapping PythonKey on all tensors, including base Tensors),
which is modeled the same way as how disabled_torch_function is done.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D34590357](https://our.internmc.facebook.com/intern/diff/D34590357)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34590357).

albanD reviewed (approved):
SGTM

zou3519 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -414,7 +416,12 @@ static PyObject* THPVariable_make_wrapper_subclass(PyObject*, PyObject* args, Py
   // forcing users to define __torch_dispatch__ we ensure this does not happen
-  TORCH_CHECK_TYPE(PyObject_FastGetAttrString(cls, "__torch_dispatch__").ptr() != nullptr,
+  // TODO: This check is not complete; because the user can disable torch
+  // dispatch and then go again, triggering segfault.  TBH I'm thinking I want
+  // to delete this function entirely
```
You want to delete _make_wrapper_subclass?

ezyang replied:
Well, I can't, because it does legitimately solve a problem that cannot be solved any other way. But it feels like there should be a better way of doing this.

ezyang commented:
@pytorchbot merge this
